cpp/include/raft/spectral/lanczos.hpp

```diff
@@ -0,0 +1,1228 @@
+
+  // Apply Lanczos method until convergence
+  shiftLower = 1;
+  shiftUpper = -1;
+  while (*totalIter < maxIter && beta_host[*effIter - 1] > tol * shiftLower) {
+    // Determine number of restart steps
+    // Number of steps must be even due to Francis algorithm
+    IndexType_ iter_new = nEigVecs + 1;
+    if (restartIter - (maxIter - *totalIter) > nEigVecs + 1)
+      iter_new = restartIter - (maxIter - *totalIter);
+    if ((restartIter - iter_new) % 2) iter_new -= 1;
+    if (iter_new == *effIter) break;
+
+    // Implicit restart of Lanczos method
+    status = lanczosRestart<IndexType_, ValueType_>(
+      handle, n, *effIter, iter_new, &shiftUpper, &shiftLower, alpha_host,
+      beta_host, Z_host, work_host, lanczosVecs_dev, work_dev, true);
+    if (status) WARNING("error in Lanczos implicit restart");
+    *effIter = iter_new;
+
+    // Check for convergence
+    if (beta_host[*effIter - 1] <= tol * fabs(shiftLower)) break;
+
+    // Proceed with Lanczos method
+    // maxIter_curr = min(restartIter, maxIter-*totalIter+*effIter);
+    status = performLanczosIteration<IndexType_, ValueType_>(
+      handle, A, effIter, maxIter_curr, *shift, tol * fabs(shiftLower),
+      reorthogonalize, alpha_host, beta_host, lanczosVecs_dev, work_dev);
+    if (status) WARNING("error in Lanczos iteration");
+    *totalIter += *effIter - iter_new;
+  }
+
+  // Warning if Lanczos has failed to converge
+  if (beta_host[*effIter - 1] > tol * fabs(shiftLower)) {
+    WARNING("implicitly restarted Lanczos failed to converge");
+  }
+
+  // Solve tridiagonal system
+  memcpy(work_host + 2 * (*effIter), alpha_host,
+         (*effIter) * sizeof(ValueType_));
+  memcpy(work_host + 3 * (*effIter), beta_host,
+         (*effIter - 1) * sizeof(ValueType_));
+  Lapack<ValueType_>::steqr('I', *effIter, work_host + 2 * (*effIter),
+                            work_host + 3 * (*effIter), Z_host, *effIter,
+                            work_host);
+
+  // Obtain desired eigenvalues by applying shift
+  for (i = 0; i < *effIter; ++i) work_host[i + 2 * (*effIter)] -= *shift;
+  for (i = *effIter; i < nEigVecs; ++i) work_host[i + 2 * (*effIter)] = 0;
+
+  // Copy results to device memory
+  CUDA_TRY(cudaMemcpy(eigVals_dev, work_host + 2 * (*effIter),
+                      nEigVecs * sizeof(ValueType_), cudaMemcpyHostToDevice));
+  // for (int i = 0; i < nEigVecs; ++i)
+  //{
+  //  std::cout <<*(work_host+(2*(*effIter)+i))<< std::endl;
+  //}
+  CUDA_TRY(cudaMemcpy(work_dev, Z_host,
+                      (*effIter) * nEigVecs * sizeof(ValueType_),
+                      cudaMemcpyHostToDevice));
+
+  // Convert eigenvectors from Lanczos basis to standard basis
+  CUBLAS_CHECK(cublasgemm(cublas_h, CUBLAS_OP_N, CUBLAS_OP_N, n, nEigVecs,
+                          *effIter, &one, lanczosVecs_dev, n, work_dev,
+                          *effIter, &zero, eigVecs_dev, n, stream));
+
+  // Clean up and exit
+  curandDestroyGenerator(randGen);
+  return 0;
```

Comment:
As mentioned above, I'd rather keep the return for a more generic interface. Let me elaborate on these reasons:

1. returning something makes the API more general; perhaps the API currently does _not_ return something but it might in the future;
2. this functions may become part of a more general interface, with different implementations (versions); these specific implementations may not return anything, but other implementations (that adhere to the same interface) might;
3. typically, this functions may return some measures (number of clusters, number of iterations, etc.); they don't currently but they might in the future; 
4. the original implementation had these return values; I'd rather err on the cautious side and assume that the original author had good reasons to return something;
5. returning a value may become useful for debugging; the return is currently trivial, but during a debug session the return could be hijacked to propagate useful information to the caller without changing the signature of the function;

Reply:
I understand the benefits and reasoning for having functions return values and I see your points above. Perhaps this promps a larger discussion about error handling & return types in RAFT. For example, we're about to move a bunch of primitives over from cuml and most if not all of these have void return types.

Reply:
In `cudf` the philosophy was to always return something (typically wrapped inside a `std::unique_ptr`). With C++1x return-value optimizations (RVO) there is no penalty in returning locals (in fact there might be performance gains, because the compiler automatically triggers the optimizers when it sees constructs like that).

So, should I `void` the above, or leave as they are (for now)?

Reply:
I'm fine leaving it like this until we have some more formal / more consistent agreement across RAFT about how we should do this. Most of the cuML prims that are in the process of being moved over have void return types, but in the absence of a formal agreement, I'm not really sure that's better than just returning 0 everywhere.